src/s3.zig

```diff
@@ -0,0 +1,2110 @@
+        this.executeSimpleS3Request(.{
+            .path = path,
+            .method = .HEAD,
+            .proxy_url = proxy_url,
+            .body = "",
+        }, .{ .stat = callback }, callback_context);
+    }
+
+    pub fn s3Download(this: *const @This(), path: []const u8, callback: *const fn (S3DownloadResult, *anyopaque) void, callback_context: *anyopaque, proxy_url: ?[]const u8) void {
+        this.executeSimpleS3Request(.{
+            .path = path,
+            .method = .GET,
+            .proxy_url = proxy_url,
+            .body = "",
+        }, .{ .download = callback }, callback_context);
+    }
+
+    pub fn s3DownloadSlice(this: *const @This(), path: []const u8, offset: usize, size: ?usize, callback: *const fn (S3DownloadResult, *anyopaque) void, callback_context: *anyopaque, proxy_url: ?[]const u8) void {
+        const range = brk: {
+            if (size) |size_| {
+                if (offset == 0) break :brk null;
+
+                var end = (offset + size_);
+                if (size_ > 0) {
+                    end -= 1;
+                }
+                break :brk std.fmt.allocPrint(bun.default_allocator, "bytes={}-{}", .{ offset, end }) catch bun.outOfMemory();
+            }
+            if (offset == 0) break :brk null;
+            break :brk std.fmt.allocPrint(bun.default_allocator, "bytes={}-", .{offset}) catch bun.outOfMemory();
+        };
+
+        this.executeSimpleS3Request(.{
+            .path = path,
+            .method = .GET,
+            .proxy_url = proxy_url,
+            .body = "",
+            .range = range,
+        }, .{ .download = callback }, callback_context);
+    }
+
+    pub fn s3StreamDownload(this: *@This(), path: []const u8, offset: usize, size: ?usize, proxy_url: ?[]const u8, callback: *const fn (chunk: bun.MutableString, has_more: bool, err: ?S3Error, *anyopaque) void, callback_context: *anyopaque) void {
+        const range = brk: {
+            if (size) |size_| {
+                if (offset == 0) break :brk null;
+
+                var end = (offset + size_);
+                if (size_ > 0) {
+                    end -= 1;
+                }
+                break :brk std.fmt.allocPrint(bun.default_allocator, "bytes={}-{}", .{ offset, end }) catch bun.outOfMemory();
+            }
+            if (offset == 0) break :brk null;
+            break :brk std.fmt.allocPrint(bun.default_allocator, "bytes={}-", .{offset}) catch bun.outOfMemory();
+        };
+
+        var result = this.signRequest(.{
+            .path = path,
+            .method = .GET,
+        }, null) catch |sign_err| {
+            if (range) |range_| bun.default_allocator.free(range_);
+
+            return switch (sign_err) {
+                error.MissingCredentials => callback(.{ .allocator = bun.default_allocator, .list = .{} }, false, .{ .code = "MissingCredentials", .message = "missing s3 credentials" }, callback_context),
```

Comment:
please use some wrapper struct that generates consistent messages from the same error codes instead of duplicating errors and messages in multiple places.